docs/RFCS/leaseholder_rebalancing.md

```diff
@@ -0,0 +1,136 @@
+to rebalance replicas away because of the restriction that we refuse
+to rebalance a replica for which a node is the leaseholder. This
+restriction is present in order to prevent an availability hiccup on
+the range when the leaseholder is removed from it.
+
+It is interesting to note a problematic behavior of the current
+system. The current leaseholder will extend its lease as long as it is
+receiving operations for a range. And when a range is split, the lease
+for the left-hand side of the split is cloned and given to the
+right-hand side of the split. The combined effect is that a newly
+created cluster that has continuous load applied against it will see a
+single node slurp up all of the range leases which causes a severe
+replica imbalance (since we can't rebalance away from the leaseholder)
+as well as a performance bottleneck. We actually see increased
+performance by periodically killing nodes in the cluster.
+
+The second motivation is to more evenly distributed load in a
+cluster. The leaseholder for a range has extra duties when compared to
+a follower: it performs all reads for a range and proposes almost all
+writes. [Proposer evaluated KV](proposer_evaluated_kv.md) will reduce
+the cost of write KV operations on followers exacerbating the
+difference between leaseholders and followers. These extra duties
+impose additional load on the leaseholder making it desirable to
+spread that load throughout a cluster in order to improve performance.
+
+The last motivation is to place the leaseholder for a range near the
+gateway node that is accessing the range in order to minimize network
+RTT. As an obvious example: it is preferable for the leaseholder to be
+in the same datacenter as the gateway node.
+
+# Detailed design
+
+This RFC is intended to address the first two motivations and punt on
+the last one (load-based leaseholder placement). Note that addressing
+the problem of evenly distributing leaseholders across a cluster will
+also address the inability to rebalance a replica away from the
+leaseholder as we'll always have sufficient non-leaseholder replicas
+in order to perform rebalancing.
+
+Leaseholder rebalancing will be performed using a similar mechanism to
+replica rebalancing. The periodically gossiped `StoreCapacity` proto
+will be extended with a `leaseholder_count` field. Currently,
+`replicateQueue` has the following logic:
+
+1. If range has dead replicas, remove them.
+2. If range is under-replicated, add a replica.
+3. If range is over-replicated, remove a replica.
+4. If the range needs rebalancing, add a replica.
+
+The proposal is to add the following logic (after the above replica
+rebalancing logic):
+
+5. If the leaseholder is on an overfull store transfer the lease to
+the least loaded follower less than the mean.
+6. If the leaseholder store has a leaseholder count above the mean and
+one of the followers has an underfull leaseholder count transfer the
+lease to the least loaded follower.
```

Comment:
This is following the same logic that we use for rebalancing replicas. If you don't have both of these rules you can get into situations where we're not rebalancing to an underfull store because other stores are above the mean but not overfull.